File: src/10.graph/46.207.ts
export function canFinish(
    numCourses: number,
    prerequisites: number[][]
): boolean {
    // 记录入度数组
    const inDegree = new Array(numCourses).fill(0);

    // 图邻接表
    const map: { [key: number]: number[] } = {};

    for (const val of prerequisites) {
        const [end, start] = val;

        // 修 end 时, 必须先修 start
        // start -> end
        // end 的入度就 + 1
        inDegree[end]++;

        // 记录邻接表
        if (map[start]) {
            map[start].push(end);
        } else {
            map[start] = [end];
        }
    }

    const queue: number[] = [];

    inDegree.forEach((item, index) => {
        // 将入度为 0 的课程加入队列
        if (item === 0) {
            queue.push(index);
        }
    });

    let count = 0;

    while (queue.length) {
        const cur = queue.shift();

        // 选课+1
        count++;

        if (map[cur]) {
            map[cur].forEach((item) => {
                inDegree[item]--;
                if (inDegree[item] == 0) {
                    queue.push(item);
                }
            });
        }
    }

    return count === numCourses;
}

File: src/10.graph/46.207.go
package graph

func CanFinish(numCourses int, prerequisites [][]int) bool {
	inDegree := make([]int, numCourses)
	m := make(map[int][]int)

	for _, v := range prerequisites {
		end, start := v[0], v[1]

		inDegree[end]++

		if _, ok := m[start]; !ok {
			m[start] = []int{end}
		} else {
			m[start] = append(m[start], end)
		}
	}

	queue := []int{}
	for i, v := range inDegree {
		if v == 0 {
			queue = append(queue, i)
		}
	}

	count := 0

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]

		if _, ok := m[cur]; ok {
			for _, v := range m[cur] {
				inDegree[v]--
				if inDegree[v] == 0 {
					queue = append(queue, v)
				}
			}
		}
		count++
	}

	return count == numCourses
}
